addon/webextension/background/communication.js

```diff
@@ -0,0 +1,67 @@
+/* globals browser, catcher */
+window.communication = (function () {
+  let exports = {};
+
+  let registeredFunctions = {};
+
+  browser.runtime.onMessage.addListener(catcher.watchFunction((req, sender, sendResponse) => {
+    if (! (req.funcName in registeredFunctions)) {
+      console.error(`Received unknown internal message type ${req.funcName}`);
+      sendResponse({type: "error", name: "Unknown message type"});
+      return;
+    }
+    if (! Array.isArray(req.args)) {
+      console.error("Received message with no .args list");
+      sendResponse({type: "error", name: "No .args"});
+      return;
+    }
+    let func = registeredFunctions[req.funcName];
+    let result;
+    try {
+      req.args.unshift(sender);
+      result = func.apply(null, req.args);
+    } catch (e) {
+      console.error(`Error in ${req.funcName}:`, e, e.stack);
+      sendResponse({type: "error", name: e+""});
+      return;
+    }
+    if (result && result.then) {
+      result.then((concreteResult) => {
+        sendResponse({type: "success", value: concreteResult});
+      }, (errorResult) => {
+        console.error(`Promise error in ${req.funcName}:`, errorResult, errorResult && errorResult.stack);
+        sendResponse({type: "error", name: errorResult+""});
+      });
+      return true;
+    } else {
+      sendResponse({type: "success", value: result});
+    }
+  }));
+
+  exports.register = function (name, func) {
+    registeredFunctions[name] = func;
+  };
+
+  /** Send a message to bootstrap.js
+      Technically any worker can listen to this.  If the bootstrap wrapper is not in place, then this
+      will *not* fail, and will return a value of exports.NO_BOOTSTRAP  */
+  exports.sendToBootstrap = function (funcName, ...args) {
+    return browser.runtime.sendMessage({funcName, args}).then((result) => {
+      if (result.type === "success") {
+        return result.value;
+      } else {
+        throw new Error(`Error in ${funcName}: ${result.name || 'unknown'}`);
+      }
+    }, (error) => {
+      if (error && error.message === "Could not establish connection. Receiving end does not exist.") {
```

Comment:
Is there another approach we could use?  I'd like to maintain the ability to run with or without bootstrap.js, but that means sending messages that might be rejected in this way.